Allow creating a branch by clicking `Enter` in the text box

tsvetilian-ty opened the pull request:
<!--
What GitHub Desktop issue does this PR address? (for example, #1234)
-->

Closes #12154

## Description

- Create a new branch on `Enter` clicked

### Screenshots
![Animation](https://user-images.githubusercontent.com/9341546/117375027-f4e7d300-aed6-11eb-9f4f-b954370992e5.gif)

tsvetilian-ty committed:
Add support for Enter clicked in search text box

cheshire137 commented on app/src/ui/lib/filter-list.tsx:
```diff
@@ -112,6 +112,9 @@ interface IFilterListProps<T extends IFilterListItem> {
   ) => void
 
+  /** Called when the Enter key is pressed in field of type search */
+  readonly onEnterClicked?: (text: string) => void
```
I could see calling this `onEnter`, `onEnterDown` (like `onKeyDown`), or `onEnterPressed`. 'Clicked' doesn't seem right for a keyboard key.

sergiou87 replied:
Agreed! I'd say we use `onEnterPressed` as we do in https://github.com/desktop/desktop/blob/28a2ee45a220dc87592e5c8e4666c4a6006b4f2b/app/src/ui/cherry-pick/choose-target-branch.tsx#L95

sergiou87 replied:
Thinking more about it, this event will be propagated not only when <kbd>Enter</kbd> is pressed, but also when the list is empty after filtering the results, so maybe we should rename it to something more generic?

Something like… `onEnterPressedWithoutFilteredItems`?

tidy-dev replied:
For more context, enter does does check out a branch when there are results. I agree with @sergiou87, clarification in the code would be nice for future devs.

cheshire137 commented on app/src/ui/lib/text-box.tsx:
```diff
@@ -271,6 +274,14 @@ export class TextBox extends React.Component<ITextBoxProps, ITextBoxState> {
+      event.key === 'Enter' &&
+      value !== undefined &&
+      value !== '' &&
+      this.props.onEnterClicked !== undefined
```
Would just checking `this.props.onEnterClicked` be sufficient without the `!== undefined`?

tidy-dev replied:
Thank you for jumping in and reviewing this! Our team has a code convention of not using falsy comparisons. Technically, it would work without the `!== undefined` but we prefer to be explicit. The only time we do that is if it is a boolean that type-wise cannot be undefined.

cheshire137 replied:
Good deal, thank you @tidy-dev!

cheshire137 commented on app/src/ui/lib/text-box.tsx:
```diff
@@ -271,6 +274,14 @@ export class TextBox extends React.Component<ITextBoxProps, ITextBoxState> {
+    } else if (
+      this.props.type === 'search' &&
+      event.key === 'Enter' &&
+      value !== undefined &&
```
Should this be `typeof value !== 'undefined'` or is that not a TypeScript thing?

sergiou87 replied:
Not a TypeScript expert, but I think that's not needed and `value !== undefined` is just fine 😄

cheshire137 reviewed (commented):
Thank you for doing this! 🙇‍♀️

tidy-dev reviewed (changes requested):
@tsvetilian-ty Thank you for working on this! I tested it out and it works great! I concur with @cheshire137's comment on the preferred naming. If you could update that, I would love to get this merged!

sergiou87 commented on app/src/ui/branches/branch-list.tsx:
```diff
@@ -81,6 +81,9 @@ interface IBranchListProps {
   /** Callback to fire when the filter text is changed */
   readonly onFilterTextChanged: (filterText: string) => void
 
+  /** Callback to fire when the filter text is changed */
+  readonly onEnterClicked?: (text: string) => void
```
I don't think we need to expose this: `BranchList` already has a method to create a branch: https://github.com/desktop/desktop/blob/28a2ee45a220dc87592e5c8e4666c4a6006b4f2b/app/src/ui/branches/branch-list.tsx#L281-L285

sergiou87 commented on app/src/ui/lib/filter-list.tsx:
```diff
@@ -372,6 +376,13 @@ export class FilterList<T extends IFilterListItem> extends React.Component<
 
+  private onEnterClicked = (text: string) => {
+    const rows = this.state.rows.length
+    if (rows === 0 && this.props.onEnterClicked) {
```
We prefer to use explicit checks for `undefined`, and I think we should also check that `text` has something in it:
```suggestion
    if (rows === 0 && text.trim().length > 0 && this.props.onEnterClicked !== undefined) {
```

tsvetilian-ty committed:
Refactor onEnterClicked in text-box

- Change onEnterClicked to onEnterPressed
- Remove unnecessary propagation of the Enter pressed event to the branch-container

tsvetilian-ty commented:
Thank you for your suggestions, everyone. 🥳

sergiou87 reviewed (dismissed):
It's looking good to me now, thanks for taking the tame to address our comments!! 🙇‍♂️ 

I also tested it and it's 😍

tsvetilian-ty committed:
fix lint issue

tidy-dev reviewed (approved):
✨  Looking good! Thank you for the updates.

jwsch51 commented:
text[](url)

1972chris reviewed (commented):
Enter
